extensions/csrc/kernel/cuda/flash_decoding_attention_kernel.cu

```diff
@@ -348,24 +351,477 @@ void flash_decoding_attention_v1_launcher(
     scale,                                                                   \
     alibi_slopes);
 
+
+template<typename scalar_t, typename cache_t, int HEAD_SIZE, int BLOCK_SIZE, int NUM_THREADS>
+__global__ void flash_decoding_attention_kernel_v2(
+  scalar_t* __restrict__ out,                 // [num_tokens, num_heads, max_num_partitions, head_size]
+  float* __restrict__ exp_sums,               // [num_tokens, num_heads, max_num_partitions]
+  float* __restrict__ max_logits,             // [num_tokens, num_heads, max_num_partitions]
+  const scalar_t* __restrict__ q,             // [num_tokens, num_heads, head_size]
+  const cache_t* __restrict__ k_cache,        // [num_blocks, num_kv_heads, head_size/x, block_size, x]
+  const cache_t* __restrict__ v_cache,        // [num_blocks, num_kv_heads, block_size, head_size]
+  const int* __restrict__ context_lens,       // [num_tokens]
+  const int* __restrict__ block_tables,       // [num_tokens, max_num_blocks_per_seq]
+  const float* __restrict__ alibi_slopes,     // [num_heads]
+  const int max_seq_len,
+  const int num_kv_heads,
+  const float scale,
+  const int max_num_blocks_per_seq,
+  const int q_stride,                         // num_heads * head_size
+  const int tmp_stride,                       // num_heads * max_num_partitions
+  const int kv_block_stride,
+  const int kv_head_stride) {
+  const int partition_idx = blockIdx.z;
+  const int seq_idx = blockIdx.y;
+  const int head_idx = blockIdx.x;
+  const int thread_idx = threadIdx.x;
+  const int lane = thread_idx % WARP_SIZE;
+  const int warp_idx = thread_idx / WARP_SIZE;
+  const int max_num_partitions = gridDim.z;
+  const int num_heads = gridDim.x;
+  const int num_queries_per_kv = num_heads / num_kv_heads;
+  const int kv_head_idx = head_idx / num_queries_per_kv;
+
+  constexpr int NUM_WARPS = NUM_THREADS / WARP_SIZE;
+  constexpr int x = sizeof(float4) / sizeof(scalar_t);
+  constexpr int Q_SHARED_SIZE = HEAD_SIZE / x;
+  // here thread_group does not determine the number of threads responsible for a key
+  // but only the VEC_SIZE of each thread
+  constexpr int THREAD_GROUP_SIZE = MAX(WARP_SIZE / BLOCK_SIZE, 1);
+  constexpr int VEC_SIZE = MIN(ROUND_DOWN_HIGHEST_POWER_OF_TWO((HEAD_SIZE / THREAD_GROUP_SIZE)), x);
```

Comment:
I wonder the relationship between ROUND_DOWN_HIGHEST_POWER_OF_TWO((HEAD_SIZE / THREAD_GROUP_SIZE)) and VEC_SIZE.

Reply:
here the data accessed by a warp should be bounded by BLOCK_SIZE x HEAD_SIZE